Assets/Scripts/Ability/AbilityManager.cs

```diff
@@ -0,0 +1,242 @@
+
+            if (_abilities[slot].CastTarget != CastTarget.Projectile)
+                UseAbilityAim(_slotNumber);
+        }
+
+        private void SetVfx()
+        {
+            for (int i = 0; i < _abilities.Count; i++)
+                _slots[i].visualEffectAsset = _abilities[i].Vfx;
+        }
+
+        private void SetupAbilities()
+        {
+            foreach (var ability in _abilities)
+                ability.Setup();
+        }
+
+        internal void ApplyEffects()
+        {
+            var ability = _abilities[_slotNumber];
+            _slotParent.transform.localPosition = ability.VfxOffset;
+
+            if (ability.CastTarget == CastTarget.Projectile)
+                UseAbilityAim(_slotNumber);
+
+            if (ability.VfxScale != Vector3.zero)
+                _slots[_slotNumber].transform.localScale = ability.VfxScale;
+
+            HandleVisualEffectTransform(_castTargetCache[_slotNumber]);
+            ExecuteAbility();
+        }
+
+        private void HandleVisualEffectTransform(CastTarget castTarget)
+        {
+            if (castTarget == CastTarget.Position)
+            {
+                _slots[_slotNumber].transform.SetParent(null, true);
+                _slots[_slotNumber].transform.position = _position;
+                _slots[_slotNumber].Reinit();
+                _slots[_slotNumber].Play();
+            }
+            else if (castTarget != CastTarget.Projectile)
+            {
+                _slots[_slotNumber].transform.SetParent(_slotParent.transform);
+                _slots[_slotNumber].transform.localPosition = Vector3.zero;
+                _slots[_slotNumber].transform.rotation = _playerCharacterTransform.rotation;
+                _slots[_slotNumber].Reinit();
+                _slots[_slotNumber].Play();
+            }
+        }
+
+        private void ExecuteAbility()
+        {
+            var ability = _abilities[_slotNumber];
+            switch (ability.CastTarget)
+            {
+                case CastTarget.Self:
+                    ability.Cast(_playerCharacterTransform.position);
+                    break;
+                case CastTarget.Direction:
+                    ability.Cast(_playerCharacterTransform.position, _playerCharacterTransform.forward);
+                    break;
+                case CastTarget.Position:
+                    ability.Cast(_position);
+                    break;
+                case CastTarget.Projectile:
+                    ability.Cast(_playerCharacterTransform.position, _playerCharacterTransform.forward, _playerCharacterTransform.rotation);
+                    break;
+                default:
+                    throw new NotImplementedException($"{nameof(CastTarget)} enum value is not implemented!");
+            }
+        }
+
+        internal void RemoveEffect()
+        {
+            _slots[_slotNumber].transform.SetParent(null, true);
+        }
+
+        public void Setup(List<Ability> abilities)
```

Comment:
Please verify that the list is not `null` and `List.count` is not 0.

Reply:
done